### ISSUE TITLE
[ANR] Reported through different file of RevenueCat/purchases-android

### ISSUE THREAD

**Issue by raghav2945** [RevenueCat/purchases-android#432]
-  [x] I have updated Purchases SDK to the latest version
-  [x] I have read the [Contribution Guidelines](https://github.com/RevenueCat/purchases-android/blob/main/CONTRIBUTING.md)
-  [x] I have searched the [Community](https://community.revenuecat.com)
-  [x] I have read [docs.revenuecat.com](https://docs.revenuecat.com/)
-  [x] I have searched for [existing Github issues](https://github.com/RevenueCat/purchases-android/issues)

**Describe the bug**
A clear and concise description of what the bug is. The more detail you can provide the faster our team will be able to triage and resolve the issue. **Do not remove any of the steps from the template below. If a step is not applicable to your issue, please leave that step empty.**

We are very pleased with your SDK. However, in recent months, we have begun to receive ANR from your SDK. Sharing the mosy frequent stack-trace we received via Firebase, as well as other relevant information in this issue.

1. Environment
   1. Platform: Android
   2. SDK version: 'com.revenuecat.purchases:purchases:4.6.1'
   3. OS version: Android 11
   4. Android Studio version: Android Studio Arctic Fox | 2020.3.1 Patch 4
   5. How widespread is the issue: Percentage of devices affected. with the range of 1000 user affected/issue
2. [Debug logs](https://docs.revenuecat.com/docs/debugging) that reproduce the issue
3. Steps to reproduce, with a description of expected vs. actual behavior
4. **Other information** (e.g. stacktraces, related issues, suggestions how to fix, links for us to have context, eg. stackoverflow, etc.)

### Stack-trace

**1. BillingWrapper.kt line 94 (com.revenuecat.purchases.google.BillingWrapper$startConnection$1.run)**

```
main (native): tid=1 systid=2145
#00 pc 0x5e2e8 libc.so 
#01 pc 0x130db3 libart.so 
#02 pc 0x4ccacb libart.so 
#03 pc 0x4cc951 libart.so 
       at android.os.BinderProxy.transactNative(Native method)
       at android.os.BinderProxy.transact(BinderProxy.java:540)
       at android.app.IActivityManager$Stub$Proxy.bindIsolatedService(IActivityManager.java:5951)
       at android.app.ContextImpl.bindServiceCommon(ContextImpl.java:1833)
       at android.app.ContextImpl.bindService(ContextImpl.java:1749)
       at android.content.ContextWrapper.bindService(ContextWrapper.java:756)
       at com.android.billingclient.api.BillingClientImpl.startConnection(com.android.billingclient:billing@@4.0.0:55)
       at com.revenuecat.purchases.google.BillingWrapper$startConnection$1.run(BillingWrapper.kt:94)
       at android.os.Handler.handleCallback(Handler.java:938)
       at android.os.Handler.dispatchMessage(Handler.java:99)
       at android.os.Looper.loop(Looper.java:223)
       at android.app.ActivityThread.main(ActivityThread.java:7700)
       at java.lang.reflect.Method.invoke(Native method)
       at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:612)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:997)
```

**2. Purchases.kt line 1854 (com.revenuecat.purchases.Purchases$Companion.configure$purchases_release)**
```
main (runnable): tid=1 systid=2975
       at kotlin.text.StringsKt__StringsJVMKt.isBlank(StringsJVM.kt:616)
       at com.revenuecat.purchases.Purchases$Companion.configure$purchases_release(Purchases.kt:1854)
       at com.revenuecat.purchases.Purchases$Companion.configure(Purchases.kt:1835)
       at com.revenuecat.purchases.Purchases$Companion.configure$default(Purchases.kt:1828)
       at com.audiomack.data.premium.PurchasesManager.<init>(PurchasesManager.kt:67)
       at com.audiomack.data.premium.PurchasesManager.<init>(PurchasesManager.kt)
       at com.audiomack.data.premium.PurchasesManager$Companion.init(PurchasesManager.kt:277)
       at com.audiomack.data.premium.PremiumRepository$Companion.init(PremiumRepository.kt:228)
       at com.audiomack.data.premium.PremiumRepository$Companion.init$default(PremiumRepository.kt:222)
       at com.audiomack.startup.PremiumInitializer.create(PremiumInitializer.kt:8)
       at com.audiomack.startup.PremiumInitializer.create(PremiumInitializer.kt:7)
       at androidx.startup.AppInitializer.doInitialize(AppInitializer.java:155)
       at androidx.startup.AppInitializer.doInitialize(AppInitializer.java:148)
       at androidx.startup.AppInitializer.discoverAndInitialize(AppInitializer.java:198)
       at androidx.startup.InitializationProvider.onCreate(InitializationProvider.java:38)
       at android.content.ContentProvider.attachInfo(ContentProvider.java:2388)
       at android.content.ContentProvider.attachInfo(ContentProvider.java:2358)
       at android.app.ActivityThread.installProvider(ActivityThread.java:7294)
       at android.app.ActivityThread.installContentProviders(ActivityThread.java:6835)
       at android.app.ActivityThread.handleBindApplication(ActivityThread.java:6730)
       at android.app.ActivityThread.access$1300(ActivityThread.java:238)
       at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1917)
       at android.os.Handler.dispatchMessage(Handler.java:106)
       at android.os.Looper.loop(Looper.java:223)
       at android.app.ActivityThread.main(ActivityThread.java:7711)
       at java.lang.reflect.Method.invoke(Native method)
       at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:952)
```

**3. PurchaserInfo.kt line 172 (com.revenuecat.purchases.PurchaserInfo.toString)**
```
main (runnable): tid=1 systid=3804
#00 pc 0x483028 libart.so 
#01 pc 0x588888 libart.so 
#02 pc 0x5a56d4 libart.so 
#03 pc 0x589564 libart.so 
#04 pc 0x6286ec libart.so 
       at java.lang.Throwable.nativeFillInStackTrace(Native method)
       at java.lang.Throwable.fillInStackTrace(Throwable.java:799)
       at java.lang.Throwable.<init>(Throwable.java:277)
       at java.lang.Exception.<init>(Exception.java:66)
       at java.lang.RuntimeException.<init>(RuntimeException.java:62)
       at java.util.MissingResourceException.<init>(MissingResourceException.java:61)
       at android.icu.util.UResourceBundle.get(UResourceBundle.java:498)
       at android.icu.impl.OlsonTimeZone.construct(OlsonTimeZone.java:504)
       at android.icu.impl.OlsonTimeZone.<init>(OlsonTimeZone.java:486)
       at android.icu.impl.ZoneMeta$SystemTimeZoneCache.createInstance(ZoneMeta.java:563)
       at android.icu.impl.ZoneMeta$SystemTimeZoneCache.createInstance(ZoneMeta.java:550)
       at android.icu.impl.SoftCache.getInstance(SoftCache.java:71)
       at android.icu.impl.ZoneMeta.getSystemTimeZone(ZoneMeta.java:580)
       at android.icu.util.TimeZone.getFrozenICUTimeZone(TimeZone.java:706)
       at android.icu.util.TimeZone.getTimeZone(TimeZone.java:687)
       at android.icu.util.TimeZone.getFrozenTimeZone(TimeZone.java:652)
       at android.icu.util.TimeZone.getDefault(TimeZone.java:857)
       at android.icu.impl.TimeZoneNamesImpl.initialize(TimeZoneNamesImpl.java:434)
       at android.icu.impl.TimeZoneNamesImpl.<init>(TimeZoneNamesImpl.java:67)
       at android.icu.impl.TimeZoneNamesFactoryImpl.getTimeZoneNames(TimeZoneNamesFactoryImpl.java:27)
       at android.icu.text.TimeZoneNames$Cache.createInstance(TimeZoneNames.java:503)
       at android.icu.text.TimeZoneNames$Cache.createInstance(TimeZoneNames.java:494)
       at android.icu.impl.SoftCache.getInstance(SoftCache.java:71)
       at android.icu.text.TimeZoneNames.getInstance(TimeZoneNames.java:157)
       at android.icu.text.TimeZoneNames.getInstance(TimeZoneNames.java:169)
       at java.util.TimeZone.getDisplayName(TimeZone.java:405)
       at java.util.Date.toString(Date.java:1066)
       at java.lang.String.valueOf(String.java:2924)
       at java.lang.StringBuilder.append(StringBuilder.java:132)
       at com.revenuecat.purchases.PurchaserInfo.toString(PurchaserInfo.kt:172)
       at java.lang.String.valueOf(String.java:2924)
       at java.lang.StringBuilder.append(StringBuilder.java:132)
       at kotlin.jvm.internal.Intrinsics.stringPlus(Intrinsics.java:20)
       at com.audiomack.data.premium.PurchasesManager.trackPurchaserInfo(PurchasesManager.kt:227)
       at com.audiomack.data.premium.PurchasesManager.processPurchaserInfo(PurchasesManager.kt:180)
       at com.audiomack.data.premium.PurchasesManager.onReceived(PurchasesManager.kt:85)
       at com.revenuecat.purchases.Purchases$sendUpdatedPurchaserInfoToDelegateIfChanged$$inlined$let$lambda$1.invoke(Purchases.kt:1479)
       at com.revenuecat.purchases.Purchases$sendUpdatedPurchaserInfoToDelegateIfChanged$$inlined$let$lambda$1.invoke(Purchases.kt:97)
       at com.revenuecat.purchases.Purchases.dispatch(Purchases.kt:1489)
       at com.revenuecat.purchases.Purchases.sendUpdatedPurchaserInfoToDelegateIfChanged(Purchases.kt:1479)
       at com.revenuecat.purchases.Purchases.afterSetListener(Purchases.kt:1462)
       at com.revenuecat.purchases.Purchases.setUpdatedPurchaserInfoListener(Purchases.kt:162)
       at com.audiomack.data.premium.PurchasesManager.<init>(PurchasesManager.kt:68)
       at com.audiomack.data.premium.PurchasesManager.<init>(PurchasesManager.kt)
       at com.audiomack.data.premium.PurchasesManager$Companion.init(PurchasesManager.kt:277)
       at com.audiomack.data.premium.PremiumRepository$Companion.init(PremiumRepository.kt:228)
       at com.audiomack.data.premium.PremiumRepository$Companion.init$default(PremiumRepository.kt:222)
       at com.audiomack.startup.PremiumInitializer.create(PremiumInitializer.kt:8)
       at com.audiomack.startup.PremiumInitializer.create(PremiumInitializer.kt:7)
       at androidx.startup.AppInitializer.doInitialize(AppInitializer.java:155)
       at androidx.startup.AppInitializer.doInitialize(AppInitializer.java:148)
       at androidx.startup.AppInitializer.discoverAndInitialize(AppInitializer.java:198)
       at androidx.startup.InitializationProvider.onCreate(InitializationProvider.java:38)
       at android.content.ContentProvider.attachInfo(ContentProvider.java:2429)
       at android.content.ContentProvider.attachInfo(ContentProvider.java:2399)
       at android.app.ActivityThread.installProvider(ActivityThread.java:8144)
       at android.app.ActivityThread.installContentProviders(ActivityThread.java:7680)
       at android.app.ActivityThread.handleBindApplication(ActivityThread.java:7500)
       at android.app.ActivityThread.access$1400(ActivityThread.java:301)
       at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2150)
       at android.os.Handler.dispatchMessage(Handler.java:106)
       at android.os.Looper.loop(Looper.java:246)
       at android.app.ActivityThread.main(ActivityThread.java:8587)
       at java.lang.reflect.Method.invoke(Native method)
       at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:602)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1130)
```

 

**Additional context**
We were unable to reproduce this issue locally, so we do not have debug logs or the steps to reproduce such an issue. It was reported via Firebase, and we shared the captured stack-trace in order to find the best solution or, at the very least, categorize these as a known issue if any.

**Comment by aboedo** [comment]
Hi @raghav2945!! 👋 Thanks for opening! 
I'll dig into this and report back

**Comment by raghav2945** [comment]
Hello @aboedo ,

Any update on this issue?

**Comment by vegaro** [comment]
@raghav2945 I am a bit confused by the stacktrace. Do you mind sharing a screenshot taken from Firebase? Thanks!

**Comment by raghav2945** [comment]
Sure no worries.
I'm sharing a screenshot of the issues that were recorded using Firebase.

![Screenshot from 2022-01-18 19-02-16](https://user-images.githubusercontent.com/4385611/149993411-b48d1055-549b-4602-87d1-2125c42d428d.png)
**Stack-trace**
```
main (native): tid=1 systid=2145
#00 pc 0x5e2e8 libc.so 
#01 pc 0x130db3 libart.so 
#02 pc 0x4ccacb libart.so 
#03 pc 0x4cc951 libart.so 
       at android.os.BinderProxy.transactNative(Native method)
       at android.os.BinderProxy.transact(BinderProxy.java:540)
       at android.app.IActivityManager$Stub$Proxy.bindIsolatedService(IActivityManager.java:5951)
       at android.app.ContextImpl.bindServiceCommon(ContextImpl.java:1833)
       at android.app.ContextImpl.bindService(ContextImpl.java:1749)
       at android.content.ContextWrapper.bindService(ContextWrapper.java:756)
       at com.android.billingclient.api.BillingClientImpl.startConnection(com.android.billingclient:billing@@4.0.0:55)
       at com.revenuecat.purchases.google.BillingWrapper$startConnection$1.run(BillingWrapper.kt:94)
       at android.os.Handler.handleCallback(Handler.java:938)
       at android.os.Handler.dispatchMessage(Handler.java:99)
       at android.os.Looper.loop(Looper.java:223)
       at android.app.ActivityThread.main(ActivityThread.java:7700)
       at java.lang.reflect.Method.invoke(Native method)
       at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:612)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:997)
```


![Screenshot from 2022-01-18 19-02-21](https://user-images.githubusercontent.com/4385611/149993416-737dc198-8933-48b7-a00b-9a22d0da806d.png)

**Stack-trace**
```
main (runnable): tid=1 systid=3804
#00 pc 0x483028 libart.so 
#01 pc 0x588888 libart.so 
#02 pc 0x5a56d4 libart.so 
#03 pc 0x589564 libart.so 
#04 pc 0x6286ec libart.so 
       at java.lang.Throwable.nativeFillInStackTrace(Native method)
       at java.lang.Throwable.fillInStackTrace(Throwable.java:799)
       at java.lang.Throwable.<init>(Throwable.java:277)
       at java.lang.Exception.<init>(Exception.java:66)
       at java.lang.RuntimeException.<init>(RuntimeException.java:62)
       at java.util.MissingResourceException.<init>(MissingResourceException.java:61)
       at android.icu.util.UResourceBundle.get(UResourceBundle.java:498)
       at android.icu.impl.OlsonTimeZone.construct(OlsonTimeZone.java:504)
       at android.icu.impl.OlsonTimeZone.<init>(OlsonTimeZone.java:486)
       at android.icu.impl.ZoneMeta$SystemTimeZoneCache.createInstance(ZoneMeta.java:563)
       at android.icu.impl.ZoneMeta$SystemTimeZoneCache.createInstance(ZoneMeta.java:550)
       at android.icu.impl.SoftCache.getInstance(SoftCache.java:71)
       at android.icu.impl.ZoneMeta.getSystemTimeZone(ZoneMeta.java:580)
       at android.icu.util.TimeZone.getFrozenICUTimeZone(TimeZone.java:706)
       at android.icu.util.TimeZone.getTimeZone(TimeZone.java:687)
       at android.icu.util.TimeZone.getFrozenTimeZone(TimeZone.java:652)
       at android.icu.util.TimeZone.getDefault(TimeZone.java:857)
       at android.icu.impl.TimeZoneNamesImpl.initialize(TimeZoneNamesImpl.java:434)
       at android.icu.impl.TimeZoneNamesImpl.<init>(TimeZoneNamesImpl.java:67)
       at android.icu.impl.TimeZoneNamesFactoryImpl.getTimeZoneNames(TimeZoneNamesFactoryImpl.java:27)
       at android.icu.text.TimeZoneNames$Cache.createInstance(TimeZoneNames.java:503)
       at android.icu.text.TimeZoneNames$Cache.createInstance(TimeZoneNames.java:494)
       at android.icu.impl.SoftCache.getInstance(SoftCache.java:71)
       at android.icu.text.TimeZoneNames.getInstance(TimeZoneNames.java:157)
       at android.icu.text.TimeZoneNames.getInstance(TimeZoneNames.java:169)
       at java.util.TimeZone.getDisplayName(TimeZone.java:405)
       at java.util.Date.toString(Date.java:1066)
       at java.lang.String.valueOf(String.java:2924)
       at java.lang.StringBuilder.append(StringBuilder.java:132)
       at com.revenuecat.purchases.PurchaserInfo.toString(PurchaserInfo.kt:172)
       at java.lang.String.valueOf(String.java:2924)
       at java.lang.StringBuilder.append(StringBuilder.java:132)
       at kotlin.jvm.internal.Intrinsics.stringPlus(Intrinsics.java:20)
       at com.audiomack.data.premium.PurchasesManager.trackPurchaserInfo(PurchasesManager.kt:227)
       at com.audiomack.data.premium.PurchasesManager.processPurchaserInfo(PurchasesManager.kt:180)
       at com.audiomack.data.premium.PurchasesManager.onReceived(PurchasesManager.kt:85)
       at com.revenuecat.purchases.Purchases$sendUpdatedPurchaserInfoToDelegateIfChanged$$inlined$let$lambda$1.invoke(Purchases.kt:1479)
       at com.revenuecat.purchases.Purchases$sendUpdatedPurchaserInfoToDelegateIfChanged$$inlined$let$lambda$1.invoke(Purchases.kt:97)
       at com.revenuecat.purchases.Purchases.dispatch(Purchases.kt:1489)
       at com.revenuecat.purchases.Purchases.sendUpdatedPurchaserInfoToDelegateIfChanged(Purchases.kt:1479)
       at com.revenuecat.purchases.Purchases.afterSetListener(Purchases.kt:1462)
       at com.revenuecat.purchases.Purchases.setUpdatedPurchaserInfoListener(Purchases.kt:162)
       at com.audiomack.data.premium.PurchasesManager.<init>(PurchasesManager.kt:68)
       at com.audiomack.data.premium.PurchasesManager.<init>(PurchasesManager.kt)
       at com.audiomack.data.premium.PurchasesManager$Companion.init(PurchasesManager.kt:277)
       at com.audiomack.data.premium.PremiumRepository$Companion.init(PremiumRepository.kt:228)
       at com.audiomack.data.premium.PremiumRepository$Companion.init$default(PremiumRepository.kt:222)
       at com.audiomack.startup.PremiumInitializer.create(PremiumInitializer.kt:8)
       at com.audiomack.startup.PremiumInitializer.create(PremiumInitializer.kt:7)
       at androidx.startup.AppInitializer.doInitialize(AppInitializer.java:155)
       at androidx.startup.AppInitializer.doInitialize(AppInitializer.java:148)
       at androidx.startup.AppInitializer.discoverAndInitialize(AppInitializer.java:198)
       at androidx.startup.InitializationProvider.onCreate(InitializationProvider.java:38)
       at android.content.ContentProvider.attachInfo(ContentProvider.java:2429)
       at android.content.ContentProvider.attachInfo(ContentProvider.java:2399)
       at android.app.ActivityThread.installProvider(ActivityThread.java:8144)
       at android.app.ActivityThread.installContentProviders(ActivityThread.java:7680)
       at android.app.ActivityThread.handleBindApplication(ActivityThread.java:7500)
       at android.app.ActivityThread.access$1400(ActivityThread.java:301)
       at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2150)
       at android.os.Handler.dispatchMessage(Handler.java:106)
       at android.os.Looper.loop(Looper.java:246)
       at android.app.ActivityThread.main(ActivityThread.java:8587)
       at java.lang.reflect.Method.invoke(Native method)
       at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:602)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1130)
```
Please let us know if there are any further details we can provide to help you fix this issue.

_The good news is that the counts appear to be reducing from the previous seven days' statistics, but the issue still persists._

**Comment by beylmk** [comment]
Hey @raghav2945 thanks for providing more detail! I've been looking into the second screenshot first, the one reported from PurchaserInfo.kt. 

PurchaserInfo.toString() will call Date.toString(), which down the line tries to apply some time zone rules. It looks like Android constructs the time zone based on the current Locale's resources, and the resources are somehow missing in this case... 

I see that you're using the App Startup jetpack library, and this ANR appears to occur during app startup. I don't have a ton of experience with that library, but is it possible that some setup order needs to be arranged? Do you customize app startup order at all? 

Those are my initial thoughts, but this is a tricky one... are you able to tell if this happens in a certain region only? 

I'll start looking into the other screenshot while I await your response :)

**Comment by raghav2945** [comment]
Hello @beylmk,

I apologize for the delay. Obtaining the location detail is difficult because it is not reported in the ANR reporting data over Firebase. As a result, determining GeoLocation is difficult. And, as previously stated, about the ANR. The numbers are decreasing over time and are restricted to previous versions of our application. It has been observed that upgrading **revenuecat.purchase** from **_4.6.0 to 4.6.1_** improves the overall health of our application. We'll keep a watch on it for the next few iterations and keep you updated, and if they're gone with the **revenuecat** upgrade, we'll be pleased to close this issue soon.

**Comment by beylmk** [comment]
Glad to hear upgrading has helped, and please do keep us posted!

**Comment by raghav2945** [comment]
Is there any news on the BillingWrapper class issue? 
This has been increasing in the last several days, and it is now back in our top five ANR problems.

Could you kindly follow up with the relevant team and let us know if anything new is on the way or if we need to make any changes to our codebase in order to use your SDK?

**Comment by beylmk** [comment]
Hey @raghav2945 sorry to hear the ANR is increasing. We created a ticket to implement a retry mechanism, but it hasn't yet been prioritized. Given that your numbers aren't improving anymore, I'll see what I can do to get that patch in soon. I'll report back here.

**Comment by beylmk** [comment]
Hey @raghav2945 just a quick update -- I've got a PR with a potential fix out, and am working through some manual and unit tests today.

I also have a quick question for you: are you implementing any kind of retry if you receive an error from any RevenueCat requests? I ask because this ANR occurs when retrying a connection to the BIllingClient too soon after it disconnects. We haven't implemented any BillingClient connection retries in our code, but we do retry the connection on each BillingClient request made. So if you were retrying those requests or if you somehow else had many back-to-back BillingClient requests, that's when this ANR would occur.

I think an exponential backoff on the reconnection is still a good idea for us to implement...I'm just trying to understand if there is something unique about your setup causing this, since we haven't had reports from other customers.

**Comment by beylmk** [comment]
Hey @raghav2945, the PR with the fix was merged! We are working on getting a release together...I'll post back here when we do.

**Comment by raghav2945** [comment]
> Hey @raghav2945, the PR with the fix was merged! We are working on getting a release together...I'll post back here when we do.

Thank you for keeping us posted.

**Comment by vegaro** [comment]
Following up on https://github.com/RevenueCat/purchases-android/issues/508

We are about to release a new version in the next couple weeks @raghav2945. We will be releasing the fix in that version.

**Comment by beylmk** [comment]
Hey @raghav2945, excited to say we have the new release out with this fix! Check out 5.0.0: https://github.com/RevenueCat/purchases-android/releases/tag/5.0.0
